server/application/application.go

```diff
@@ -369,7 +369,13 @@ func (s *Server) Create(ctx context.Context, q *application.ApplicationCreateReq
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "unable to check existing application details (%s): %v", appNs, err)
 	}
-	equalSpecs := reflect.DeepEqual(existing.Spec, a.Spec) &&
+
+	if err := argo.ValidateDestination(ctx, &existing.Spec.Destination, s.db); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "application destination spec for %s is invalid: %s", existing.Name, err.Error())
+	}
+
+	equalSpecs := existing.Spec.Destination.Equals(a.Spec.Destination) &&
+		reflect.DeepEqual(existing.Spec, a.Spec) &&
```

Comment:
> if the user does change from name to server in the spec that's still a valid change to the spec and we should be updating the resource in the cluster?

True, but as any other change done to the app spec it needs an upsert since the spec was changed, isn't it?

> Plus as far as I can see in your code if there's already an app with just destination name set then ValidateDestination() will also add the server component to the actual spec of the App so if the user sends the exact same app with the name set the DeepEqual check will fail because your ValidateDestination just update the spec with the missing field effectively making it not equal spec even though it was equal spec

This is not true since the `ValidateDestination()` is also called for the application from the request via `validateAndNormalizeApp` function call, see line 337 above.